### PR TITLE
8343239: Javadoc marks non-preview type as preview if it's the subtype of a preview type

### PR DESCRIPTION
**Pull request by nizarbenalla**
Please review this patch to improve the javadoc preview header. The parent bug include more details.

This is meant to prevent a bug where, if a non preview class extend an `@PreviewFeature` interface, the Javadoc displays `Programs can only use <<Type>> when preview features are enabled` when the API can be used without enable preview features as long as we don't refer to any methods in the supertype

This is specifically the case in  #17543. Here is the javadoc output [AsymmetricKey](https://cr.openjdk.org/~nbenalla/GeneratedDocs/8343239-2/docs/api/java.base/java/security/AsymmetricKey.html) after merging this branch with the PR `298420: PEM API: Implementation (Preview) #17543`. The only files affected by this change are the class `DEREncodable` and it's Subinterfaces.

One other bug I noticed is that, on the `DEREncodable` page, it's direct subinterfaces had the superscript `PREVIEW` link. I've added a condition to remove it.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343239](https://bugs.openjdk.org/browse/JDK-8343239): Javadoc marks non-preview type as preview if it's the subtype of a preview type (**Bug** - P4)

### Reviewers
 * [Hannes Wallnöfer](https://openjdk.org/census#hannesw) (@hns - **Reviewer**) Review applies to [f8b3110c](https://git.openjdk.org/jdk/pull/22126/files/f8b3110c389d65c6cddbafbb28f554aeb93ebab1)
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22126/head:pull/22126` \
`$ git checkout pull/22126`

Update a local copy of the PR: \
`$ git checkout pull/22126` \
`$ git pull https://git.openjdk.org/jdk.git pull/22126/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22126`

View PR using the GUI difftool: \
`$ git pr show -t 22126`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22126.diff">https://git.openjdk.org/jdk/pull/22126.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22126#issuecomment-2478612468)
</details>

### CODE REVIEW

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties [hunk 1]
```diff
@@ -439,6 +439,8 @@ Programs can only use <code>requires transitive java.base</code> when \
 preview features are enabled.<br>\
 Preview features may be removed in a future release, or upgraded \
 to permanent features of the Java Platform.<br>
+doclet.NonPreviewPreview=Programs can use {0} without enabling preview features. \
+    Programs can only reference the preview API by name when preview features are enabled.
```
Comment (liach) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties:443]: > Programs can only reference the preview API by name when preview features are enabled.

Should we reword like `Programs can only use {0} as {1} when preview features are enabled.` where `{1}` is (the list of) preview feature class(es) or interface(s) this existing class or interface extends/implements.

**Comment by nizarbenalla** [comment]
Slightly confusing to pull changes from the PEM PR and then remove them before pushing. I have improved the preview message on the non-preview interfaces.

**Comment by hns** [comment]
I'm not sure the non-preview class needs to get a preview notice at all just for implementing an (empty) preview interface. I think somebody with more knowledge about preview features should have a look at this.

**Comment by nizarbenalla** [comment]
AFAICT [this case](https://github.com/openjdk/jdk/blob/9769ee86978584a65703712c1a845fe4f5fffe29/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java#L2520-L2542) is not handled by any test, removing this block only broke `test/langtools/jdk/javadoc/tool/CheckResourceKeys.java` which is a more general check doesn't directly check if the output is correct for custom code. (I checked manually and the output is the same after my changes if custom classes extend a `@PreviewFeature` class)

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 2]
```diff
@@ -2515,7 +2515,8 @@ public void addPreviewInfo(Element forWhat, Content target) {
             Content note2 = contents.getContent("doclet.PreviewTrailingNote2", nameCode);
             previewDiv.add(HtmlTree.DIV(HtmlStyles.previewComment, note2));
             target.add(previewDiv);
-        } else if (forWhat.getKind().isClass() || forWhat.getKind().isInterface()) {
+        } else if ((forWhat.getKind().isClass() || forWhat.getKind().isInterface())
+                && !utils.nonPreviewExtendsPreview(forWhat)) {
```
[the remark this answers is not in the document]
Comment (nizarbenalla) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java]: I think I fixed it in [f8b3110](https://github.com/openjdk/jdk/pull/22126/commits/f8b3110c389d65c6cddbafbb28f554aeb93ebab1). The solution looks simpler without any of the added methods, thanks.

> (An interface extending a preview interface is a different case again, although I don't think that should ever occur in the JDK.)

Is this a typo? because I think this is exactly what the [PEM Encodings JEP](https://github.com/openjdk/jdk/pull/17543/files) does.

```
public non-sealed interface AsymmetricKey extends Key, DEREncodable {
}
```

**Review by hns — APPROVED** [review]
Looks good to me, and it solves the problem with our given use case.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java [hunk 3]
```diff
@@ -2485,7 +2485,6 @@ public PreviewSummary declaredUsingPreviewAPIs(Element el) {
                     usedInDeclaration.addAll(types2Classes(tpe.getBounds()));
                 }
                 usedInDeclaration.addAll(types2Classes(List.of(te.getSuperclass())));
-                usedInDeclaration.addAll(types2Classes(te.getInterfaces()));
                 usedInDeclaration.addAll(types2Classes(te.getPermittedSubclasses()));
```
Comment (liach) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/util/Utils.java:2488]: This one is dubious too - imagine if we have a classfile `Attribute` that is preview-only, such as `Pattern` or `LoadableDescriptors`.  But that is out of scope for this patch.

File: test/langtools/jdk/javadoc/doclet/testPreview/api2/nonpreviewextendspreview/CoreInterface.java
Comment (liach) [inline test/langtools/jdk/javadoc/doclet/testPreview/api2/nonpreviewextendspreview/CoreInterface.java:1]: These files are tiny, yet they bring in extra directory structures and huge license headers.  Can we try to use `ToolBox.writeJavaFiles` to generate these source files from text blocks instead?

**Review by liach — CHANGES_REQUESTED** [review]
I don't think we should add 3 new files when we can just use `ToolBox.writeJavaFiles`.

**Comment by nizarbenalla** [comment]
I had to resolve a merge conflict after [JDK-8318416](https://bugs.openjdk.org/browse/JDK-8318416), I've also made changes based on chen's suggestions.

**Review by liach — APPROVED** [review]
Looks very neat; thanks!

**Comment by nizarbenalla** [comment]
Thank you for the reviews! I'm glad the final solution ended up being very simple.

/integrate
